core/src/main/java/org/elasticsearch/index/percolator/ExtractQueryTermsService.java

```diff
@@ -170,6 +179,35 @@ public static void extractQueryTerms(Query query, ParseContext.Document document
         } else if (query instanceof BlendedTermQuery) {
             List<Term> terms = ((BlendedTermQuery) query).getTerms();
             return new HashSet<>(terms);
+        } else if (query instanceof SpanTermQuery) {
+            return Collections.singleton(((SpanTermQuery) query).getTerm());
+        } else if (query instanceof SpanNearQuery) {
+            Set<Term> bestClause = null;
+            SpanNearQuery spanNearQuery = (SpanNearQuery) query;
+            for (SpanQuery clause : spanNearQuery.getClauses()) {
+                Set<Term> temp = extractQueryTerms(clause);
+                bestClause = selectTermListWithTheLongestShortestTerm(temp, bestClause);
+            }
+            return bestClause;
+        } else if (query instanceof SpanOrQuery) {
+            Set<Term> terms = new HashSet<>();
+            SpanOrQuery spanOrQuery = (SpanOrQuery) query;
+            for (SpanQuery clause : spanOrQuery.getClauses()) {
+                terms.addAll(extractQueryTerms(clause));
+            }
+            return terms;
+        } else if (query instanceof SpanFirstQuery) {
+            return extractQueryTerms(((SpanFirstQuery)query).getMatch());
+        } else if (query instanceof SpanNotQuery) {
+            return extractQueryTerms(((SpanNotQuery) query).getInclude());
+        } else if (query instanceof FieldMaskingSpanQuery) {
```

Comment:
maybe we should not have optimizations for FieldMaskingSpanQuery, it looks dangerous?